Fixed a bug that filling NULL bytes when changing the attribute after renaming

ggtakec opened the pull request:
## Relevant Issue (if applicable)
#1453 

### Details
The rename logic internally copies the file and deletes the original file.
The cause of the #1453 was that s3fs was updating mtime immediately after this copy, at which time a new cache file was created and the contents of the file were null.

In this PR, updates such as mtime are added to the cache file before copying.
And leave the cache file open and copy (put_header).
After copying, rename the cache file and delete the original file.

Keeping the cache file open is necessary because fdcache cannot be renamed unless it is open and new cache files are not created.

And, although it was also the cause of the problem, I accidentally deleted the cache of the destination file.
This is also fixed. (Prevents unnecessary downloads and improves efficiency)

gaul commented:
Does this fix the failing test in #1564?  I still see this on my system:

```
test-s3fs.txt /dev/fd/63 differ: byte 1, line 1
test_update_metadata_external_object failed
```

ggtakec commented:
@gaul Please review new fixed code.
The reason for the failure of the test being used in #1564 was due to a code other than `rename`.
However, the root cause is related to `put_header` and a similar fix is needed.
Fixed the PR code and fixed a similar bug with `chmod`/`chown`/`utimens`.
Also, since the #1564 test had only one pattern of chmod, I added a test to test `chmod`/`chown`/`utimens` with a `small file` and a `large file`.
The new modified code confirms that these tests have passed.
